### ISSUE TITLE
A (root) window's "safe area" method that works regardless of window position, device, screen id, notch, rounded corners, etc

### ISSUE THREAD

**Issue by geekley** [godotengine/godot-proposals#11106]
### Describe the project you are working on

A cross-platform 2D game that needs to work on Web, Mobile and PC, on any device.

### Describe the problem or limitation you are having in your project

Having to adapt to unsafe display areas like cutouts and rounded corners is difficult, specially when you don't have the devices to properly test different arrangements. For example, I know some phones have rounded corners, but no notch, but I don't have any such device available for testing.

I need certain UI elements to be at screen corners, and I don't want to push away by an arbitrary amount on every device (even the ones with normal rect screen) as it would ruin my layout.

My current solution that uses margins based on just `DisplayServer.get_display_cutouts()` won't be enough, even though that method works regardless of window position (Android apps can be put in floating or side-by-side windows), because presumably that method doesn't consider rounded corners (I can't test to be sure).

There's also `DisplayServer.get_display_safe_area()` and `DisplayServer.screen_get_usable_rect()`, but the are for the screen, not app window, and on a PC I can already see one of them excludes things like the taskbar of the current screen. Which means I would probably have to burn some neurons trying to figure out how to account for screen position -- of course, even assuming that works for rounded corners, which is not clear in the documentation -- and all without having devices to actually test if my custom solution would even work...

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be so much better if you could add a method on the root window like `root.get_safe_area()` or even better `root.unsafe_margins.[top|right|left|bottom]` that would already work on any device regardless of window position, screen id, etc. It must be the full size of the window if it's unobstructed, but be eaten up by margins if there's any notch, rounded corners or whatever in the way ***within the window***.

Then app designers don't need to try figuring out this stuff and coming up with their own custom solutions which will end up not really working on all devices that they can't really test properly. If this is done by Godot, then people with different devices can test better, and eventual updates and fixes will work for everyone.

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

Layouts that need to be pushed away from cutouts, notches, rounded corners, etc. could add a MarginContainer with a script that applies margins based on the result of the `root_window.unsafe_margins.*` properties. You don't have to worry about window position, screen id or device OS, it would just work.

Note that this example would be only for the top / full rect control node that expands across the entire window. The ideal would be if it's possible to somehow improve it to work on any sub rect within the app.

A simpler alternative might also be a project setting that already makes "full-screen" not consider notch/rounded/system areas, so you're guaranteed a free rectangle to work with. But whatever solution should work on mobile browsers too.

```gdscript
class_name SafetyMarginContainer extends MarginContainer

@onready var _root_window: Window = get_tree().root

func _process(_delta: float) -> void:
	add_theme_constant_override(&'margin_left', _root_window.unsafe_margins.left)
	add_theme_constant_override(&'margin_top', _root_window.unsafe_margins.top)
	add_theme_constant_override(&'margin_right', _root_window.unsafe_margins.right)
	add_theme_constant_override(&'margin_bottom', _root_window.unsafe_margins.bottom)

```

### If this enhancement will not be used often, can it be worked around with a few lines of script?

I do have a script, but it works only for cutouts. Probably won't do rounded corners (can't test), and trying to improve it to use `DisplayServer.get_display_safe_area()` or `DisplayServer.screen_get_usable_rect()` would require thought on how to deal with window position, and test on multiple devices that I don't have.

I could only test this on Android phones (one normal unobstructed rect and one that has both camera cutout and rounded corners) in a single screen (no external 2nd display), and it just works also on floating window and side-by-side windows.

<details><summary>CutoutsMargin script to apply margins. Use as the top (full rect) control, having other nodes under it.</summary>

```gdscript
# Public domain - Unlicense

class_name CutoutsMargin extends MarginContainer
## Sets the margins to avoid display cutouts, like notches, cameras and sensors over the screen.

## Force horizontal symmetry by applying the larger margin to both sides.
@export var symmetric_horizontally: bool = true
## Force vertical symmetry by applying the larger margin to both sides.
@export var symmetric_vertically: bool = true

func _enter_tree() -> void:
	set_margins_from_cutouts()

func _process(_delta: float) -> void:
	set_margins_from_cutouts()

var display_cutouts: Array[Rect2] = []
func set_margins_from_cutouts() -> void:
	var cutouts: Array[Rect2] = DisplayServer.get_display_cutouts()
	if cutouts == display_cutouts: return
	display_cutouts = cutouts
	var screen_size: Vector2i = DisplayServer.screen_get_size()
	var margin_lt: Vector2i
	var margin_rb: Vector2i
	for cutout: Rect2 in cutouts:
		var cutout_center: Vector2 = cutout.get_center()
		if cutout_center.x < screen_size.x / 3.0:
			margin_lt.x = maxi(margin_lt.x, roundi(cutout.end.x))
		elif cutout_center.x > screen_size.x * 2 / 3.0:
			margin_rb.x += maxi(margin_rb.x, roundi(screen_size.x - cutout.position.x))
		if cutout_center.y < screen_size.y / 3.0:
			margin_lt.y += maxi(margin_lt.y, roundi(cutout.end.y))
		elif cutout_center.y > screen_size.y * 2 / 3.0:
			margin_rb.y += maxi(margin_rb.y, roundi(screen_size.y - cutout.position.y))
	if symmetric_horizontally:
		var margin_h: int = maxi(margin_lt.x, margin_rb.x)
		margin_lt.x = margin_h
		margin_rb.x = margin_h
	if symmetric_vertically:
		var margin_v: int = maxi(margin_lt.y, margin_rb.y)
		margin_lt.y = margin_v
		margin_rb.y = margin_v
	add_theme_constant_override(&'margin_left', margin_lt.x)
	add_theme_constant_override(&'margin_top', margin_lt.y)
	add_theme_constant_override(&'margin_right', margin_rb.x)
	add_theme_constant_override(&'margin_bottom', margin_rb.y)

```

</details>

### Is there a reason why this should be core and not an add-on in the asset library?

It's an improvement over the `DisplayServer` methods. Everyone doing cross-platform games would benefit from this.
In fact, you could even add a more advanced `SafetyMarginContainer` node to core that already applies this logic, to improve layouts with a feature to push away from unsafe screen areas automatically, regardless of node nesting level.

It would be so much better if game designers can just focus on working on their game layout, without having to worry about custom scripts to deal with specific problematic device features, that many would probably not even have considered.

**Comment by danijmn** [comment]
I agree, every mobile game developer would benefit from this added convenience.

But IMO even more importantly, it should be seen as an opportunity to fix issues in the current implementation, such as godotengine/godot#105462, godotengine/godot#55463, godotengine/godot#99809 and godotengine/godot-proposals#3466.
